examples/TestRoutingGuard/can-load.spec.ts

```diff
@@ -0,0 +1,172 @@
+const canLoadGuard: CanLoadFn = (route, segments) => {
+  if (route && segments && inject(LoginService).isLoggedIn) {
+    return true;
+  }
+
+  return from(inject(Router).navigate(['/login'])).pipe(mapTo(false));
+};
+
+// Another canLoad guard like in a real world example,
+// which should be removed from testing to avoid side effects on the route.
+const sideEffectCanLoadGuard: CanLoadFn = () => false;
+
+// A simple component pretending a login form.
+// It will be replaced with a mock copy.
+@Component({
+  selector: 'login',
+  template: 'login',
+})
+class LoginComponent {}
+
+// A simple component pretending a protected dashboard.
+// It will be replaced with a mock copy.
+@Component({
+  selector: 'dashboard',
+  template: 'dashboard',
+})
+class DashboardComponent {}
+
+@NgModule({
+  declarations: [DashboardComponent],
+  imports: [
+    RouterModule.forChild([
+      {
+        path: 'dashboard',
+        component: DashboardComponent,
+      },
+      {
+        path: '**',
+        redirectTo: 'dashboard',
+      },
+    ]),
+  ],
+  exports: [],
+})
+class DashboardModule {}
+
+// Definition of the routing module.
+@NgModule({
+  declarations: [LoginComponent, DashboardComponent],
+  exports: [RouterModule],
+  imports: [
+    RouterModule.forRoot([
+      {
+        component: LoginComponent,
+        path: 'login',
+      },
+      {
+        canLoad: [canLoadGuard, sideEffectCanLoadGuard],
+        path: '',
+        loadChildren: () => DashboardModule,
+      },
+    ]),
+  ],
+  providers: [LoginService],
+})
+class TargetModule {}
+
+describe('TestRoutingGuard:canLoad', () => {
+  // Because we want to test a canLoad guard, it means that we want to
+  // test its integration with RouterModule.
+  // Therefore, RouterModule and the guard should be kept,
+  // and the rest of the module which defines the route can be mocked.
+  // To configure RouterModule for the test,
+  // RouterModule, RouterTestingModule.withRoutes([]), NG_MOCKS_ROOT_PROVIDERS
+  // should be specified as the first parameter of MockBuilder (yes, with empty routes).
+  // The module with routes and the guard should be specified
+  // as the second parameter of MockBuilder.
+  // Then `NG_MOCKS_GUARDS` should be excluded to remove all guards,
+  // and `canLoadGuard` should be kept to let you test it.
+  beforeEach(() => {
+    return MockBuilder(
+      [
+        RouterModule,
+        RouterTestingModule.withRoutes([]),
+        NG_MOCKS_ROOT_PROVIDERS,
+      ],
+      [TargetModule],
+    )
+      .exclude(NG_MOCKS_GUARDS)
+      .keep(canLoadGuard);
+  });
+
+  it('redirects to login', async () => {
+    if (Number.parseInt(VERSION.major, 10) < 7) {
+      pending('Need Angular  7+');
+
+      return;
+    }
+
+    const fixture = MockRender(RouterOutlet, {});
+    const router = ngMocks.get(Router);
+    const location = ngMocks.get(Location);
+
+    // First we need to initialize navigation.
+    if (fixture.ngZone) {
```

Comment:
Starting with Angular now comes with the option to be zoneless.
Does this spec function under this condition?
Could we trigger the navigation in another way, so to not rely on the ngZone?

Reply:
You raised a valid point here. 
I would rather consider creating a new issue to check, if everything is still working on a zoneless application. The way how the test is implemented here, is also used in other specs. 
Probably it would make sense to have e2e tests for Angular 18 Zoneless. But as I said therefore a new issue should be created.

Reply:
@satanTime 
From looking into the past stuff, I saw that there where two branches at one point because of the different rendering engines.

Do you think this could be an approach here?
Having 2 branches per version, after Angular 18, one for zoneless and another with it?

I'm aware from reading the angular team blogs and github that they want to move away from it, but not sure if it will actually occur, or when. Using the two branches approaches is the safest I can think of.

Reply:
it's just a test, and, for now, it's fine that it depends on `zone.js`, because `ng-mocks` examples include this library by default.

It doesn't block usage of `ng-mocks` and `canLoad` in projects without `zone.js`.